### PR TITLE
Add ability to have unique table location for each delta lake table

### PR DESCRIPTION
**Pull request by homar**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

a connector

> How would you describe this change to a non-technical end user or system administrator?

each table will have different location random location, unique suffix will be added to the table name

## Related issues, pull requests, and links

https://github.com/trinodb/trino/issues/12980

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Delta Lake
* Add ability to have unique table location for each delta lake table  ({[issue](https://github.com/trinodb/trino/issues/12980)}`12980`)
```

### CODE REVIEW

**Comment by homar** [comment]
Failure is the same as here https://starburstdata.atlassian.net/browse/SEP-9405

**Comment by findepi** [comment]
is the ci failure related?

**Comment by homar** [comment]
> is the ci failure related?

no it is not

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
@alexjo2144 @ebyhr ptal

> no it is not

if it's a flake, is there an issue already?

**Comment by ebyhr** [comment]
I believe the failed test was already removed in https://github.com/trinodb/trino/commit/308e717baf78dc344d7f962bf32155416b751902.

**Comment by ebyhr** [comment]
I believe the test was already removed in https://github.com/trinodb/trino/commit/308e717baf78dc344d7f962bf32155416b751902.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeTableWithCustomLocation.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.List;
+import java.util.Optional;
+
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+public abstract class TestDeltaLakeTableWithCustomLocation
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeTableWithCustomLocation.java:37]: It would be better to rename to `BaseDeltaLakeTableWithCustomLocation`.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeTableWithCustomLocation.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        extends AbstractTestQueryFramework
+{
+    protected static final String SCHEMA = "tables_with_custom_location";
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeTableWithCustomLocation.java:40]: If I remember correctly, cleanup test for Glue expects `test_` prefix for the schema name.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeTableWithCustomLocation.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    protected static final String CATALOG_NAME = "delta_with_custom_location";
+    protected File metastoreDir;
+    protected HiveMetastore metastore;
+    protected HdfsEnvironment hdfsEnvironment;
+    protected HdfsEnvironment.HdfsContext hdfsContext;
+
+    @Test
+    public void testTableHasUuidSuffixInLocation()
+    {
+        String tableName = "table_with_uuid";
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeTableWithCustomLocation.java:50]: What happens if Glue test is executed at the same time? Don't we need to use random suffix?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeTableWithCustomLocation.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        assertQuerySucceeds(format("CREATE TABLE %s as select 1 as val", tableName));
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeTableWithCustomLocation.java:51]: nit: Uppercase some keyword
```suggestion
        assertQuerySucceeds(format("CREATE TABLE %s AS SELECT 1 AS val", tableName));
```

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeTableWithCustomLocationUsingGlueMetastore.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+        queryRunner.installPlugin(new DeltaLakePlugin());
+        queryRunner.createCatalog(
+                CATALOG_NAME,
+                CONNECTOR_NAME,
+                ImmutableMap.<String, String>builder()
+                        .put("hive.metastore", "glue")
+                        .put("hive.metastore.glue.region", "us-east-2")
+                        .put("hive.metastore.glue.default-warehouse-dir", metastoreDir.getPath())
+                        .put("delta.hive-catalog-name", "hive_with_redirections")
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeTableWithCustomLocationUsingGlueMetastore.java:69]: I suppose this config property is unnecessary in this test.

**Comment by homar** [comment]
@ebyhr comments applied

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeTableWithCustomLocation.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+public abstract class BaseDeltaLakeTableWithCustomLocation
+        extends AbstractTestQueryFramework
+{
+    protected static final String SCHEMA = "test_tables_with_custom_location";
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeTableWithCustomLocation.java:41]: Can we add `randomTableSuffix()` considering Glue?

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeTableWithCustomLocation.java (hunk 6 above)
Comment (homar) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeTableWithCustomLocation.java:41]: @ebyhr do you mean to the schema name ?

**Comment by ebyhr** [comment]
Rebased on upstream to resolved conflicts.

**Comment by homar** [comment]
@ebyhr can we merge this ?

**Comment by ebyhr** [comment]
@homar `TestDeltaLakeAdlsConnectorSmokeTest#testManagedTableFilesCleanedOnDrop` looks failing consistently (https://github.com/trinodb/trino/runs/7481219750 & in my laptop). Could you take a look at that? 

```
Error:  io.trino.plugin.deltalake.TestDeltaLakeAdlsConnectorSmokeTest.testManagedTableFilesCleanedOnDrop  Time elapsed: 1.97 s  <<< FAILURE!
java.lang.AssertionError: 
Expecting:
 <0>
to be greater than:
 <0> 
	at io.trino.plugin.deltalake.BaseDeltaLakeConnectorSmokeTest.testManagedTableFilesCleanedOnDrop(BaseDeltaLakeConnectorSmokeTest.java:713)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by homar** [comment]
@ebyhr I pushed a fix for ADSL, could you please run ADSL tests again ?

**Comment by ebyhr** [comment]
Sure, updated #13310

**Comment by homar** [comment]
@ebyhr I think now are good to merge, if my last change to fix that adsl test is fine for you

**Review by findepi — APPROVED** [review]

**Comment by ebyhr** [comment]
Merged, thanks!

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 7]
```diff
@@ -1605,7 +1616,9 @@ private void finishOptimize(ConnectorSession session, DeltaLakeTableExecuteHandl
 
     private boolean allowWrite(ConnectorSession session, DeltaLakeTableHandle tableHandle)
     {
-        boolean requiresOptIn = transactionLogWriterFactory.newWriter(session, tableHandle.getLocation()).isUnsafe();
+        String tableLocation = metastore.getTableLocation(tableHandle.getSchemaTableName(), session);
+        Path tableMetadataDirectory = new Path(new Path(tableLocation).getParent().toString(), tableHandle.getTableName());
+        boolean requiresOptIn = transactionLogWriterFactory.newWriter(session, tableMetadataDirectory.toString()).isUnsafe();
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1621]: Why was this change needed?
